kbase/narrative
# DATAUP-487 Staging Area Viewer reworking to allow uploading of all checked files

# Description of PR purpose/changes

This has a number of changes to the staging area viewer code to support some bug fixes. There's a few other changes that this included.

Sorry for the size, there was really no way to do this piecemeal, as it needed an overhaul of how the datatable gets created and maintained. I couldn't really see a way to do that in parts without introducing strange and unusual behavior.

Bugs addressed:
* Navigating back from a two-deep-nested directory used to go from, for example `user/dir1/dir2` to `user/user/dir1`. Now it goes to `user/dir1` as expected.
* Selected files from multiple pages now all get added to the import cell.
* Select all / select none button works again.
* Selected files from nested directories all get added to the import cell.
  * For example, selecting a file from root, a file from one subdirectory, and files from a deeper directory, will all get added to the bulk import cell
* File checked state persists through refreshes.

Structural / architecture changes:
* The `render` function should now only get run once. This builds the DataTable, which should also only be done once.
* Previously, we used data built into the DOM with a template file and applied the DataTable to that. Now, we just feed the data rows into it.
* The DataTable now knows how to render the data from each file object. On refresh, this fetches the data again from the server, and just feeds it into the DataTable.
* This now fetches all the files from the Staging Area and just feeds them to the table, which filters based on what's the current directory. This really isn't that much of a change, as the `StagingArea.list()` function worked recursively. So running `list()` on the root directory returned all files anyway. We now just take advantage of that instead of throwing things away.
* Each file is now treated as unique based on its path, not its file name. If there are files with the same name in multiple directories, this should now be ok - before it led to Strange Behavior.

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-487

# Testing Instructions
* Details for how to test the PR:
* Throw some files in your staging area, enough to show multiple pages / directories
* Toggle through pages and select files
* Do the import step, and watch them all appear!
* Also, try a refresh, deletion, download, or other function to make sure nothing was lost

# Dev Checklist:
